File: python/ocr-numbers/ocr_numbers.py
"""
Convert a visual representation of numbers in a list of strings to 
a string of those numbers.
"""

DIGIT_WIDTH = 3
DIGIT_HEIGHT = 4

digits = [
 '    _  _     _  _  _  _  _  _ ', #
 '  | _| _||_||_ |_   ||_||_|| |', # Decimal numbers.
 '  ||_  _|  | _||_|  ||_| _||_|', #
 '                              ' # The fourth line is always blank
]

def convert(input_grid):
    """
    Convert a list of strings to the numbers they represent.
    :param input_grid: list of strings with OCR-able characters
    :returns: String of numbers represented by the input grid.
    """
    if len(input_grid) == 0:
        return ''
    
    # when the rows aren't multiples of 4
    if len(input_grid) % 4 != 0:
        raise ValueError('Number of input lines is not a multiple of four')

    # when the columns aren't multiples of 3
    for row in input_grid:
        if len(row) % 3 != 0 or len(row) == 0:
            raise ValueError('Number of input columns is not a multiple of three')
    ret = ''
    for row_index in range(0,len(input_grid), DIGIT_HEIGHT):
        for col_index in range(0, len(input_grid[row_index]), DIGIT_WIDTH):
            data = [input_grid[index][col_index:(col_index + DIGIT_WIDTH)] for index in range(row_index, row_index + DIGIT_HEIGHT)]
            ret += match_digit(data)
        ret += ','
    return ret[0:len(ret) - 1]

def match_digit(data):
    """Check if a given set of data matches a digit pattern.
    :param data: list of DIGIT_HEIGHT rows each a string of DIGIT_WIDTH characters to check against known digit patterns.
    :returns: '?' if no match found, otherwise the matching digit converted to a single character string.
    """
    for digit in range(10):
        match = (digit + 1) % 10
        for row_index in range(DIGIT_HEIGHT):
            row = digits[row_index][digit * DIGIT_WIDTH: (digit + 1) * DIGIT_WIDTH]
            if data[row_index] != row:
                match = -1
                break
        if match >= 0:
            return str(match)
    return '?'

File: go/ocr-numbers/ocr_numbers.go
// OCR Numbers exercise. Really just string matching.
package ocrnumbers

import (
	"errors"
	"strconv"
	"strings"
)

// Rearranged to put zero first so index matches digit.
var digits []string = []string{
	" _     _  _     _  _  _  _  _ ", //
	"| |  | _| _||_||_ |_   ||_||_|", // Decimal numbers.
	"|_|  ||_  _|  | _||_|  ||_| _|", //
	"                              ", // The fourth line is always blank,
}

// Recognize a string of digits from the given \n separated string.
// @param s: Input text (string)
// @returns: slice of strings where ocr digits are the expected number (? if not recognized). One entry for matching block.
// @raises: Error if the array is not the correct size/shape.
func Recognize(s string) ([]string, error) {
	result := make([]string, 0)
	rows := strings.Split(s, "\n")
	if len(rows) > 0 && len(rows[0]) == 0 {
		rows = rows[1:]
	}
	if len(rows) < 4 || len(rows)%4 != 0 {
		return result, errors.New("number of input lines is not a multiple of four")
	}
	for i := 1; i < len(rows); i++ {
		if len(rows[i]) < 3 || len(rows[i])%3 != 0 {
			return result, errors.New("number of input columns is not a multiple of three")
		}
	}

	for j := 0; j < len(rows); j += 4 {
		var number string = ""
		for i := 0; i < len(rows[j]); i += 3 {
			digit, err := recognizeDigit(rows, i, j)
			if err != nil {
				number += "?"
			} else {
				number += strconv.Itoa(digit)
			}
		}
		result = append(result, number)
	}
	return result, nil
}

// Recognize a single digit from the slice of string at position x, y
// @param rows: slice of string one string per line in the input
// @param x: x-coordinate of the position to match at.
// @param y: y-coordinate of the position to match at.
// @returns: The digit that was matched, or an error.
// @raises: Raises and error if no digit match can be found.
func recognizeDigit(rows []string, x int, y int) (int, error) {
	for i := 0; i < len(digits[0]); i += 3 {
		if rows[y+0][x:x+3] == digits[0][i:i+3] &&
			rows[y+1][x:x+3] == digits[1][i:i+3] &&
			rows[y+2][x:x+3] == digits[2][i:i+3] &&
			rows[y+3][x:x+3] == digits[3][i:i+3] {
			return i / 3, nil
		}
	}
	return 0, errors.New("No match found.")
}
